docs/14.EXT-REFERENCE-REG-FUNCTIONS.md

```diff
@@ -0,0 +1,184 @@
+# New struct definitions
+
+## jerryx_function_list_entry
+
+**Summary**
+
+Structure to define an array of methods with `name` and `function` fields which
+can be registered to a target object. When setting the `function` value
+a native C method must be used.
+
+**Prototype**
+
+```c
+typedef struct {
+  const char *name;
+  jerry_external_handler_t function;
+} jerryx_function_list_entry;
```

Comment:
My only problem with this list is that if you have a list for registering functions, later you will need a list for registering objects, strings, symbols numbers as well. Since the list contains not more than a key-value pairs I think all the mentioned types can be handled in one list. Then the structure would look like somehow like this:

```c
typedef struct {
  const char* name;
  void* data;
  jerry_value_type_t type;
}
```

I think it would be really useful to have a single universal structure for the API values registration what do you think about it?

Reply:
The idea is good. However, there is (at least) one problem, take the following example:
```c
struct {
  const char* name;
  void* data;
  jerry_value_type_t type;
} entries[] = {
 { "name", 3.2, JERRY_TYPE_NUMBER},
 ...
};
```
In this case the user would want to store the `3.2` as a double value which is a 64 bit data. However, on a 32 bit system the `void*` has a size of 32 bit thus we can not put a 64 bit data in it.

Another thing I was wondering is this:
```c
struct {
  const char* name;
  jerry_type_t type;
  union {
    double number;
    void* function;
    char* str;
    bool boolean;
    jerry_value_t object;
  }
} entries[] = {
  { "name", JERRY_TYPE_NUMBER, { .number = 3.2 } },
 ..
};
```
However this form of union initialization can not be used in C++ ( if I'm correct the C++17 standard or C++20 supports this now).

So.... I don't see a universally usable method allowing all kind of property value registration :(.

Reply:
However, I think we can use a workaround for the `double` variables.

```c
double my_double_variable = 3.2;

struct {
  const char* name;
  void* data;
  jerry_value_type_t type;
} entries[] = {
 { "name", &my_double_variable, JERRY_TYPE_NUMBER},
 ...
};
```

I think it could work on 32 and 64 bit as well. The only requirement is to this solution is that the double variable must be declared and initialized before the it gets part of the initialization list. The only disadvantage is that `JERRY_TYPE_NUMBER` could not be used since integer variables can be passed by value, but IMO we can introduce a new enum for the property registration. What do you think about it?

Reply:
That could work, yes. I'm not sure how useful it is in practice to have an extra variable for doubles. In addition if we go this way the initialization of the entries should be the same, thus all other data should be also initialized the same way (eg.: jerry_value, bool, ) to have a common interface.

Reply:
Okay that was only a suggestion, here come the second:

```c
/* entry registration for non-number types */
#define JERRY_EXT_GENERAL_ENTRY(name, value, type) \
  { name, (void *) value, type}

/* entry registration for number types */
#define JERRY_EXT_NUMBER_ENTRY(name, value) \
  { name, (void *) (jerry_create_number (value)), JERRY_TYPE_NUMBER}

struct {
  const char* name;
  void* data;
  jerry_value_type_t type;
} entries[] = {
 JERRY_EXT_GENERAL_ENTRY ("foo1", "string value", JERRY_TYPE_STRING),
 JERRY_EXT_GENERAL_ENTRY ("foo2", true, JERRY_TYPE_BOOLEAN),
 JERRY_EXT_NUMBER_ENTRY ("foo3", 4),
 JERRY_EXT_NUMBER_ENTRY ("foo4", 3.2),
 ...
};
```

Truly the number registration is a bit problematic, but I think it's acceptable to use an other macro for it. What's your opinion about that?

Reply:
Going further we could even use `jerry_value_t` as the data type and expect the user to correctly create all values and in addition we'll provide the macros.

Reply:
I support this idea!